File: vehicle_simulator/src/transform_cloud_node.cpp
#include <ros/ros.h>
#include <nav_msgs/Odometry.h>
#include <sensor_msgs/PointCloud2.h>
#include <sensor_msgs/point_cloud2_iterator.h>
#include <geometry_msgs/Pose.h>
#include <geometry_msgs/Quaternion.h>
#include <std_msgs/Header.h>
#include <tf/transform_datatypes.h>
#include <Eigen/Dense>
#include <vector>

class PointCloudTransformer
{
public:
  PointCloudTransformer(ros::NodeHandle &nh)
      : nh_(nh)
  {
    // std::string ns;
    // nh_.param<std::string>("robot_namespace", ns, "vehicle0");

    // pose_topic_ = ns + "/state_estimation_noisy";
    // scan_topic_ = ns + "/sensor_scan";
    // output_topic_ = ns + "/transformed_cloud";

    pose_topic_ = "state_estimation_noisy";
    scan_topic_ = "sensor_scan";
    output_topic_ = "transformed_cloud";

    pose_sub_ = nh_.subscribe(pose_topic_, 10, &PointCloudTransformer::poseCallback, this);
    cloud_sub_ = nh_.subscribe(scan_topic_, 10, &PointCloudTransformer::cloudCallback, this);
    cloud_pub_ = nh_.advertise<sensor_msgs::PointCloud2>(output_topic_, 10);

    latest_pose_received_ = false;
    // ROS_INFO("PointCloud Transformer Initialized for [%s]", ns.c_str());
  }

private:
  ros::NodeHandle nh_;
  ros::Subscriber pose_sub_, cloud_sub_;
  ros::Publisher cloud_pub_;
  std::string pose_topic_, scan_topic_, output_topic_;

  geometry_msgs::Pose latest_pose_;
  bool latest_pose_received_;

  void poseCallback(const nav_msgs::Odometry::ConstPtr &msg)
  {
    latest_pose_ = msg->pose.pose;
    latest_pose_received_ = true;
  }

  void cloudCallback(const sensor_msgs::PointCloud2::ConstPtr &msg)
  {
    if (!latest_pose_received_)
    {
      ROS_WARN_THROTTLE(5.0, "No pose received yet.");
      return;
    }

    Eigen::Affine3d transform = getTransformMatrix(latest_pose_);
    std::vector<Eigen::Vector3f> transformed_points;

    sensor_msgs::PointCloud2ConstIterator<float> iter_x(*msg, "x");
    sensor_msgs::PointCloud2ConstIterator<float> iter_y(*msg, "y");
    sensor_msgs::PointCloud2ConstIterator<float> iter_z(*msg, "z");

    for (; iter_x != iter_x.end(); ++iter_x, ++iter_y, ++iter_z)
    {
      Eigen::Vector4d pt(*iter_x, *iter_y, *iter_z, 1.0);
      Eigen::Vector4d pt_map = transform.matrix() * pt;
      transformed_points.emplace_back(pt_map[0], pt_map[1], pt_map[2]);
    }

    sensor_msgs::PointCloud2 output;
    output.header.stamp = msg->header.stamp;
    output.header.frame_id = "map";
    output.height = 1;
    output.width = transformed_points.size();
    output.is_dense = false;
    output.is_bigendian = false;
    sensor_msgs::PointCloud2Modifier modifier(output);
    modifier.setPointCloud2FieldsByString(1, "xyz");
    modifier.resize(transformed_points.size());

    sensor_msgs::PointCloud2Iterator<float> out_x(output, "x");
    sensor_msgs::PointCloud2Iterator<float> out_y(output, "y");
    sensor_msgs::PointCloud2Iterator<float> out_z(output, "z");

    for (size_t i = 0; i < transformed_points.size(); ++i, ++out_x, ++out_y, ++out_z)
    {
      *out_x = transformed_points[i].x();
      *out_y = transformed_points[i].y();
      *out_z = transformed_points[i].z();
    }

    cloud_pub_.publish(output);
  }

  Eigen::Affine3d getTransformMatrix(const geometry_msgs::Pose &pose)
  {
    Eigen::Translation3d translation(pose.position.x, pose.position.y, pose.position.z);
    Eigen::Quaterniond q(pose.orientation.w, pose.orientation.x, pose.orientation.y, pose.orientation.z);
    return translation * q;
  }
};

int main(int argc, char **argv)
{
  ros::init(argc, argv, "pointcloud_transformer");
  ros::NodeHandle nh;
  PointCloudTransformer transformer(nh);
  ros::spin();
  return 0;
}

File: vehicle_simulator/scripts/transform_cloud_node.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
Author: lee lizw_0304@163.com
Date: 2025-07-08 11:23:26
LastEditors: lee lizw_0304@163.com
LastEditTime: 2025-07-08 11:30:00
FilePath: /AUTO_DEVELOPMENT/vehicle_simulator/scripts/transform_cloud_node.py
Description: 订阅车辆位姿和原始点云，将点云转换到 map 坐标系下发布
"""

import rospy
import tf
import numpy as np
from nav_msgs.msg import Odometry  # ✅ 正确的消息类型
from sensor_msgs.msg import PointCloud2
import sensor_msgs.point_cloud2 as pc2
from std_msgs.msg import Header


class PointCloudTransformer:
    def __init__(self):
        rospy.init_node("pointcloud_transformer")

        self.robot_namespace = rospy.get_param("~robot_namespace", "vehicle0")
        self.pose_topic = "/" + self.robot_namespace + "/state_estimation_noisy"
        self.scan_topic = "/" + self.robot_namespace + "/sensor_scan"

        self.latest_pose = None
        self.tf_listener = tf.TransformListener()

        # ✅ 订阅 nav_msgs/Odometry
        rospy.Subscriber(self.pose_topic, Odometry, self.pose_callback)
        rospy.Subscriber(self.scan_topic, PointCloud2, self.scan_callback)

        self.pub = rospy.Publisher(
            "/" + self.robot_namespace + "/transformed_cloud", PointCloud2, queue_size=1
        )
        rospy.loginfo(
            "PointCloud Transformer Initialized for [%s]" % self.robot_namespace
        )

    def pose_callback(self, msg):
        # ✅ 提取 pose.pose，而不是 PoseStamped
        self.latest_pose = msg.pose.pose

    def scan_callback(self, cloud_msg):
        if self.latest_pose is None:
            rospy.logwarn_throttle(5.0, "No pose received yet.")
            return

        # 提取位姿
        position = self.latest_pose.position
        orientation = self.latest_pose.orientation
        trans = (position.x, position.y, position.z)
        rot = (orientation.x, orientation.y, orientation.z, orientation.w)

        # 构造 4x4 的变换矩阵
        T = self.get_transform_matrix(trans, rot)

        # 点云转换
        points = list(
            pc2.read_points(cloud_msg, field_names=("x", "y", "z"), skip_nans=True)
        )
        transformed_points = []
        for p in points:
            pt = np.array([p[0], p[1], p[2], 1.0])
            pt_map = T.dot(pt)
            transformed_points.append((pt_map[0], pt_map[1], pt_map[2]))

        # 构造新的 PointCloud2 消息
        header = Header()
        header.stamp = cloud_msg.header.stamp
        header.frame_id = "map"  # 可根据需要改为 odom 或 base_link

        cloud_out = pc2.create_cloud_xyz32(header, transformed_points)
        self.pub.publish(cloud_out)

    def get_transform_matrix(self, translation, quaternion):
        tf_matrix = tf.transformations.quaternion_matrix(quaternion)
        tf_matrix[0:3, 3] = translation
        return tf_matrix


if __name__ == "__main__":
    try:
        transformer = PointCloudTransformer()
        rospy.spin()
    except rospy.ROSInterruptException:
        pass
